core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,129 @@
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* encode.c -- an encoder for DR synthetic IR */
+
+#include "encode.h"
+
+#include "../globals.h"
+#include "encode_api.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Encodes DR instruction representation \p instr into raw bytes \p encoded_instr.
+ * Returns next instruction's pc.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
+ */
+byte *
+encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
+{
+    /* Use a local uint variable for easier setting of category, eflags, #src, and #dst
+     * values.
+     */
+    uint encoding_header = 0;
+
+    /* Encode number of register destination operands (i.e., written registers).
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands (i.e., read registers).
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     */
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
+    uint eflags = 0;
+    if (TESTANY(EFLAGS_WRITE_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_WRITES_ARITH;
+    if (TESTANY(EFLAGS_READ_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_READS_ARITH;
+    encoding_header |= (eflags << FLAGS_SHIFT);
+
+    /* Encode category as synthetic opcode.
+     */
+    uint category = instr_get_category(instr);
+    encoding_header |= (category << CATEGORY_SHIFT);
+
+    /* Copy header encoding back into encoded_instr output.
+     */
+    *((uint *)&encoded_instr[0]) = encoding_header;
+
+    /* Encode register destination operands, if present.
+     */
+    opnd_size_t max_opnd_size = OPSZ_NA;
+    for (uint dst_index = 0; dst_index < num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr, dst_index);
+        max_opnd_size = opnd_get_size(dst_opnd);
```

Comment:
The max is not properly updated here; nor below.

Reply:
It doesn't need to, because they are all the same.
Although let's discuss why that's the case.

Reply:
Not needed anymore, as discussed offline.